pylsp/hookspecs.py

```diff
@@ -80,12 +80,12 @@ def pylsp_folding_range(config, workspace, document):
 
 
 @hookspec(firstresult=True)
-def pylsp_format_document(config, workspace, document):
+def pylsp_format_document(config, workspace, document, options):
```

Comment:
Unfortunately, this is not the right way of passing options to plugins. Plugins read options from the `config` object passed as first parameter in this call.

You can see how that's handled by (for instance) the pycodestyle plugin here:

https://github.com/python-lsp/python-lsp-server/blob/develop/pylsp/plugins/pycodestyle_lint.py

So, what you need to do is basically something similar for Yapf.

Reply:
These are options passed to the request, not configuration options set ahead of time. Unless I'm misunderstanding

Reply:
Right, but all our plugins work like this. It makes little sense to expect that every request has its own set of configuration options because users can't control that on their editors/IDEs.

Instead, clients are expected to send a `workspace/didChangeConfiguration` request with a new set of options, those are saved in our config system, and then the next formatting or linting request will read and use them.

Reply:
The `textDocument/formatting` request has explicit formatting options, the configuration isn't something that is specified in the protocol, it's specified as `LSPAny`. There isn't a well-known way to specify formatting options in the workspace configuration. To add, not all the files follow the same formatting options so this isn't really a workspace-wide config.

> all our plugins work like this

Which plugins/requests are you referring to exactly?

> It makes little sense to expect that every request has its own set of configuration options because users can't control that on their editors/IDEs.

How so? Client implementations send formatting options based on user's options.

Reply:
I think that @ccordoba12 meant workspace-wide settings as in:

https://github.com/python-lsp/python-lsp-server/blob/f19b3b9d44f4a2d5ab1e241f83a3fbc3390d25cc/pylsp/plugins/jedi_completion.py#L38-L47

But I agree with you @masad-frost here. For `textDocument/formatting` request the LSP protocol defines `DocumentFormattingParams`  which includes `options: FormattingOptions;`.

How IDEs decide to pass this options is irrelevant - they may hard-code the options, or expose them to user per-workspace or per-document - their choice.

Reply:
> For textDocument/formatting request the LSP protocol defines DocumentFormattingParams which includes options: FormattingOptions

You're right. I didn't know the LSP allows to pass options to this specific request.

Reply:
Oh, thanks @krassowski for clarifying the misunderstanding, maybe my PR description wasn't clear.

I'm currently traveling but can probably fix up the PR's tests in the coming week. Let me know if there's other comments. 🙏